clusterloader2/pkg/measurement/util/wait_for_pods.go

```diff
@@ -49,31 +54,69 @@ type WaitForPodOptions struct {
 // WaitForPods waits till desired number of pods is running.
 // Pods are be specified by namespace, field and/or label selectors.
 // If stopCh is closed before all pods are running, the error will be returned.
-func WaitForPods(clientSet clientset.Interface, stopCh <-chan struct{}, options *WaitForPodOptions) error {
-	ps, err := NewPodStore(clientSet, options.Selector)
-	if err != nil {
-		return fmt.Errorf("pod store creation error: %v", err)
+func WaitForPods(clientSet clientset.Interface, stopCh <-chan struct{}, options *WaitForPodOptions) (finalErr error) {
+	var lister func() []*corev1.Pod
+	if options.Selector.FieldSelector == "" && options.Lister != nil {
```

Comment:
I like this change, but I'm concerned about cpu resource usage growth of clusterloader.

We have done a test with this PR cherry picked and it looks like the CPU usage grows a lot:

* Without this PR:
![image](https://user-images.githubusercontent.com/2559168/161980758-63814c23-7546-4dab-86b9-8131f3f308de.png)

* With this PR:
![image](https://user-images.githubusercontent.com/2559168/161980843-18d496d6-8c7c-42ba-b189-d641b26b1a5c.png)

This make sense as now, for each deployment being updated we will be scanning all pods (in a given namespace) and filter pods manually, every 5s. Before this PR, this load was done by kube-apiserver (which is also bad).

Can we optimize this somehow?

Reply:
We can try to use indexers like in https://github.com/kubernetes/kubernetes/blob/4e4a868e138ca1ecff7d126859f01822db1b5ee1/pkg/controller/daemon/daemon_controller.go#L190-L192

but this would require changing WaitForControlledPodsRunning API to ask user for label to index by. If not provided, fallback to old watch per deployment logic.

Reply:
I'm not sure if or when I'll have time for such an improvement. I could take this part out of the PR and submit it separately, then perhaps the rest can be merged.

Reply:
> We can try to use indexers like in https://github.com/kubernetes/kubernetes/blob/4e4a868e138ca1ecff7d126859f01822db1b5ee1/pkg/controller/daemon/daemon_controller.go#L190-L192

> but this would require changing WaitForControlledPodsRunning API to ask user for label to index by. If not provided, fallback to old watch per deployment logic.


In practise, I don't think we need to over-optimize it here. I believe that having a static index by namespace should be enough.
here is my proposal on how to do that: https://github.com/kubernetes/perf-tests/pull/2056